lib/matplotlib/backends/_backend_tk.py

```diff
@@ -198,14 +199,33 @@ def __init__(self, figure=None, master=None):
         # event to the window containing the canvas instead.
         # See https://wiki.tcl-lang.org/3893 (mousewheel) for details
         root = self._tkcanvas.winfo_toplevel()
-        root.bind("<MouseWheel>", self.scroll_event_windows, "+")
+
+        # Prevent long-lived references via tkinter callback structure GH-24820
+        weakself = weakref.ref(self)
+        weakroot = weakref.ref(root)
+
+        def scroll_event_windows(event):
+            self = weakself()
+            if self is None:
+                root = weakroot()
+                if root is not None:
+                    root.unbind("<MouseWheel>", scroll_event_windows_id)
```

Comment:
I guess I'm a little hung up on how this cleanup gets done/namespacing here...

if `scroll_event_windows_id` is not itself garbage collected, then how would `self` be garbage collected (and thus the weakref returns None), but root is not garbage collected (and thus that weakref is not None)

while the `self` from the outer namespace is occluded inside of this function by the first line, variables from that namespace are available inside of this function, and thus I'm not sure how the strong references to those variables can be dead yet ever?

and if that is not how the namespacing works, and thus the variables from the outer namespace can be gc'd, then `scroll_event_windows_id` may itself be gone (though shenanigans with integers and the GC are always hard to track)

Its admittedly been a while since I thought hard about weakref lifetimes and when the namespaces die for internally defined functions... 

Almost seems like you would have to define this outside of the namespace where `self` and `root` are directly and pass them in with a `functools.partial` or something (along with the unbind id)

Do you have empirical data to show that this does, in fact, prevent the memory build up?

I'm willing to accept it as not too crazy and weakrefs doing their job, but just raised some questions for me about whether the lifetimes are extended by the way this function is defined (and thus all the weakrefs are just for show).

Reply:
Empirically, I can say this does help, using the test from #24820, the rate of growth of memory usage as a function of button presses is lower (not gone, but lower)

To put numbers on it on my system, the example script uses ~83 MB at startup.
After 20 clicks of the "update graph" button on main, I get ~130 MB memory usage (~1.5x).
With this patch applied, I get ~104 MB memory usage after the same number of clicks.


Okay, I experimented with some simpler examples of creating these kinds of closures just to convince myself of how it works... and yes, indeed it only captures variables _used_ and thus I see how this would work, keep a strong ref to the id (which is probably fine) but weak refs to self and root.

```python
>>> def outer(x):
...    def inner():
...      print(locals().keys())
...    return inner
... 
>>> outer(1)()
dict_keys([])
>>> def outer(x):
...    def inner():
...      print(locals().keys())
...      print(x)
...    return inner
... 
>>> outer(1)()
dict_keys(['x'])
1
```

Reply:
I totally understand the confusion here. I'm no weakref expert myself.

> if `scroll_event_windows_id` is not itself garbage collected, then how would `self` be garbage collected (and thus the weakref returns None), but root is not garbage collected (and thus that weakref is not None)

This is precisely the case if a canvas is deleted but the window remains active
 
> while the `self` from the outer namespace is occluded inside of this function by the first line, variables from that namespace are available inside of this function, and thus I'm not sure how the strong references to those variables can be dead yet ever?

I specifically chose to shadow the names "root" and "self" in the callback namespaces so that it is difficult to accidentally reference them in a closure

> and if that is not how the namespacing works, and thus the variables from the outer namespace can be gc'd, then `scroll_event_windows_id` may itself be gone (though shenanigans with integers and the GC are always hard to track)

`scroll_event_windows_id` is a closure from the local namespace of `self.__init__` but the object itself is a string and so does not have references to `self` or anything else for that matter, so there's no cycles to worry about.

> Its admittedly been a while since I thought hard about weakref lifetimes and when the namespaces die for internally defined functions...

python closes over anything that it needs, even if it might seem like it should be dying... that's our whole problem here!

> Do you have empirical data to show that this does, in fact, prevent the memory build up?

I was seeing better than 90% reduction in growth after ensuring that `last_event` was released and `calling gc.collect()`

> I'm willing to accept it as not too crazy and weakrefs doing their job, but just raised some questions for me about whether the lifetimes are extended by the way this function is defined (and thus all the weakrefs are just for show).

This is a good argument for maybe having some more comments.